### PR TITLE
Create/human scoring/download scorer

### PR DESCRIPTION
**Pull request by sirpa44**

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -4,7 +4,9 @@
     "require": {
         "php": "^7.2",
         "ext-ctype": "*",
+        "ext-http": "*",
```
Comment (peetya) [inline composer.json]: What is the purpose of this line? Is this an existing extension?

File: README.md [hunk 2]
```diff
@@ -54,6 +54,46 @@ Before running the tests please make sure you have PHP SQLite extension installe
...
+###settings
+
+in 
+```bash
+config/services.yaml
+```
+set the adapter.
+```bash
+App\Command\Adapter\ProviderInterface:
+        alias: App\Command\Adapter\<Adapter>
```
Comment (peetya) [inline README.md]: The adapter is already added to `services.yaml` so it's not required to add to the README.

File: src/Command/ImportScorerCommand.php [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+     *
+     * @param InputInterface $input
+     * @param OutputInterface $output
+     * @throws \Exception
+     */
+    protected function initialize(InputInterface $input, OutputInterface $output)
+    {
+        $this->input = $input;
+        $this->output = $output;
+        $this->stopwatch->start('import');
```
Comment (peetya) [inline src/Command/ImportScorerCommand.php]: You would get more accurate measurement if you start the stopwatch in the execution phase (`execute()` method). What do you think?

File: src/Command/ImportScorerCommand.php [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Console\Style\SymfonyStyle;
+use Symfony\Component\Security\Core\Encoder\UserPasswordEncoderInterface;
+use Symfony\Component\Stopwatch\Stopwatch;
+
+class ImportScorerCommand extends Command
+{
+    CONST OVERWRITE = '--overwrite';
+    CONST FORCE = '--force';
```
Comment (peetya) [inline src/Command/ImportScorerCommand.php]: `const` should be lowercase. Also it would make sense to change the name of them to something more accurate, like:
```
const OPTION_OVERWRITE = 'overwrite';
const OPTION_FORCE = 'force';
const OPTION_PATH = 'path';
```

... as you are referring to these texts in many places.

What do you think?

File: src/Command/ImportScorerCommand.php [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+     * @param InputInterface $input
+     * @param OutputInterface $output
+     * @return int
+     */
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+
+        $this->symfonyStyle->title('Human-Scoring Import');
+        if (!$this->input->hasParameterOption(self::OVERWRITE)) {
+            $this->output->writeln(['if you want overwrite scorer data in database use flag \'--overwrite\'', '']);
```
Comment (peetya) [inline src/Command/ImportScorerCommand.php]: Since the `input` and the `output` parameters are provided to the `execute()` function, it would make sense to use them instead of the class variables.

File: src/Command/ImportScorerCommand.php [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+     * convert and show StopWatch data
+     *
+     * @param $memory
+     * @param $duration
+     */
+    private function showStopwatchData($memory, $duration)
+    {
+        $data['memory'] = round($memory / 1000000, 2);
+        $data['duration'] = round($duration / 1000, 2);
+        $this->output->writeln(['memory: ' . $data['memory'] . ' Mo' , 'duration: ' . $data['duration'] . ' second']);
```
Comment (peetya) [inline src/Command/ImportScorerCommand.php]: `MB` instead of `Mo` and `seconds` instead of `seconds`.

File: src/Command/Adapter/Csv.php [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+     *
+     * @param $filePath
+     * @param $symfonyStyle
+     * @return array
+     */
+    public function getIterator($filePath, $symfonyStyle)
+    {
+        $fileExtension = pathinfo($filePath, PATHINFO_EXTENSION);
+        if ($fileExtension !== 'csv' || !file_exists($filePath)) {
+            $symfonyStyle->warning('Data file path incorrect.');
```
Comment (peetya) [inline src/Command/Adapter/Csv.php]: It's not a good approach to mix the purpose of the classes. The purpose of this class is to provide an Adapter for CSV files, so the `symfonyStyle` shouldn't be passed to this method.
You should throw only the exceptions and catch them in your Command and display an error message depending on them.

File: src/Command/Adapter/Csv.php [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+use Symfony\Component\HttpFoundation\File\Exception\FileNotFoundException;
+
+class Csv implements ProviderInterface
+{
+    /**
+     * return an array of Scorer data from a csv file
+     *
+     * @param $filePath
+     * @param $symfonyStyle
+     * @return array
```
Comment (peetya) [inline src/Command/Adapter/Csv.php]: It will return with an `Iterator` and not array.

File: src/Command/Adapter/Csv.php [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+     */
+    public function getIterator($filePath, $symfonyStyle)
+    {
+        $fileExtension = pathinfo($filePath, PATHINFO_EXTENSION);
+        if ($fileExtension !== 'csv' || !file_exists($filePath)) {
+            $symfonyStyle->warning('Data file path incorrect.');
+            throw new FileNotFoundException('Data file path incorrect.');
+        }
+        $reader = Reader::createFromPath($filePath, 'r');
```
Comment (peetya) [inline src/Command/Adapter/Csv.php]: `'r'` parameter is not required as it is the default value.

File: src/Command/Adapter/ProviderInterface.php [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+/**
+ * Human Scoring Software
+ *
+ * @author antoinep@taotesting.com
+ * @license See LICENCE.md
+ */
+namespace App\Command\Adapter;
+
+
+interface ProviderInterface
```
Comment (peetya) [inline src/Command/Adapter/ProviderInterface.php]: It should be more like `AdapterInterface`. What do you think?

File: src/Command/Adapter/ProviderInterface.php [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+{
+    /**
+     * return an array of Scorers data
+     *
+     * @param $filePath
+     * @param $symfonyStyle
+     * @return mixed
```
Comment (peetya) [inline src/Command/Adapter/ProviderInterface.php]: Would be great if you could define the type of the @params and also the @return should be Iterator.

File: src/Command/ImportScorerCommand.php [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+            'force',
+            'f',
+            InputOption::VALUE_NONE,
+            'Causes data ingestion to be applied into storage'
+        );
+        $this->addOption(
+          'overwrite',
+            'o',
+            InputOption::VALUE_NONE,
+            'overwrite Scorer in database'
```
Comment (peetya) [inline src/Command/ImportScorerCommand.php:88]: I think it's not required to have both parameters. The goal is to run the command in `dry-run` mode which means that it get executed on a regular way without saving anything into the database and have the `--force` attribute which will the persisting.
It doesn't make sense to run this command without the `--overwrite` attribute as nothing will happen.

**Review by peetya — CHANGES_REQUESTED** [review]

File: README.md [hunk 13]
```diff
@@ -54,6 +54,46 @@ Before running the tests please make sure you have PHP SQLite extension installe
 $ ./bin/phpunit
 ```
 
+## Import Scorer Command
+
+### CSV File
```
Comment (siwane) [inline README.md:87]: Please could u provide an explicit example

File: README.md [hunk 14]
```diff
@@ -54,6 +54,46 @@ Before running the tests please make sure you have PHP SQLite extension installe
...
+```bash
+--force
+```
+Causes data ingestion to be applied into storage
+
+Overwrite :
+```bash
+--overwrite
+```
+overwrite Scorer in database
```
Comment (siwane) [inline README.md]: Could u explain this difference between `overwrite` & `force`, it is non human readable

File: README.md [hunk 15]
```diff
@@ -54,6 +54,46 @@ Before running the tests please make sure you have PHP SQLite extension installe
 $ ./bin/phpunit
 ```
 
+## Import Scorer Command
```
Comment (siwane) [inline README.md:83]: Could u explain what is this command please?

File: src/Command/Adapter/Csv.php (hunk 7 above)
Comment (siwane) [reply src/Command/Adapter/Csv.php]: :+1:

**Review by siwane — CHANGES_REQUESTED** [review]
the command does not work with both options when data exists. Also the command is not intuitive (helper for overwrite is displayed only when is applicable), please add a detailed report to command to inform user about data existing / number of added / number of updated ...

File: README.md [hunk 16]
```diff
@@ -54,6 +54,42 @@ Before running the tests please make sure you have PHP SQLite extension installe
...
+--force
+```
+Import Scorer Command is a DryRun by default.
+force option causes data ingestion to be applied into storage.
+
+Overwrite :
+```bash
+--overwrite
+```
+Import Scorer Command does't change scorer data for a Scorer that already exist in database by default.
```
Comment (peetya) [inline README.md]: typo: `does't` instead of `doesn't`

File: config/services.yaml [hunk 17]
```diff
@@ -25,3 +25,11 @@ services:
 
     # add more service definitions when explicit configuration is needed
     # please note that last definitions always *replace* previous ones
+#
+
+    App\Repository\ScorerRepository:
+        tags: ['doctrine.repository_service']
```
Comment (peetya) [inline config/services.yaml]: Why is it required to tag it as a repository service?

File: config/services.yaml [hunk 18]
```diff
@@ -25,3 +25,11 @@ services:
...
+
+
+    App\Command\Adapter\ProviderInterface:
```
Comment (peetya) [inline config/services.yaml]: This class is not exists.

File: data/scorer.csv [hunk 19]
```diff
@@ -0,0 +1,3 @@
+username,password
```
Comment (peetya) [inline data/scorer.csv]: Please mark this file as an example, like for example moving it from `data/scorer.csv` to `data/examples/scorers.csv`.

File: src/Command/Adapter/AdapterInterface.php [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+namespace App\Command\Adapter;
+
+
+interface AdapterInterface
+{
+    /**
+     * return an array of Scorers data
+     *
+     * @param $filePath
+     * @return iterator
```
Comment (peetya) [inline src/Command/Adapter/AdapterInterface.php]: `iterator` is not defined. I think you wanted to add `\Iterator`.

File: src/Command/Adapter/AdapterInterface.php [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+     */
+    public function getIterator($filePath);
```
Comment (peetya) [inline src/Command/Adapter/AdapterInterface.php:19]: Please add response type declaration.

File: src/Command/Adapter/Csv.php [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+
+class Csv implements AdapterInterface
+{
+    /**
+     * return an array of Scorer data from a csv file
+     *
+     * @param $filePath
+     * @return iterator
+     */
+    public function getIterator($filePath)
```
Comment (peetya) [inline src/Command/Adapter/Csv.php:21]: `\League\Csv\Exception` exception is not handled.

File: src/Command/ImportScorerCommand.php [hunk 23]
```diff
@@ -0,0 +1,184 @@
...
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Console\Style\SymfonyStyle;
+use Symfony\Component\Security\Core\Encoder\UserPasswordEncoderInterface;
+use Symfony\Component\Stopwatch\Stopwatch;
+
+class ImportScorerCommand extends Command
+{
+    const OPTION_OVERWRITE = '--overwrite';
+    const OPTION_FORCE = '--force';
+    const ARGUMENT_PATH = 'path';
```
Comment (peetya) [inline src/Command/ImportScorerCommand.php:27]: Please add access modifiers for these contants. (I would say `private`).

File: src/Command/ImportScorerCommand.php [hunk 24]
```diff
@@ -0,0 +1,184 @@
...
+            $data = $this->dataProvider->getIterator($this->path);
+            $this->persist($data);
+            $this->flush();
+        } catch (\Exception $e) {
+            echo $e->getMessage();
+            return 1;
+        }
+            $event = $this->stopwatch->stop('import');
+            $this->showStopwatchData($event->getMemory(),$event->getDuration ());
+            return 0;
```
Comment (peetya) [inline src/Command/ImportScorerCommand.php:139]: Wrong indentation for these 3 lines.

File: src/Command/ImportScorerCommand.php [hunk 25]
```diff
@@ -0,0 +1,184 @@
...
+    }
+
+
+    /**
+     * if there is option '--overwrite' persist
+     * persist scorers
+     *
+     * @param $data
```
Comment (peetya) [inline src/Command/ImportScorerCommand.php:147]: Type declaration for `$data` parameter is missing.

File: src/Command/ImportScorerCommand.php [hunk 26]
```diff
@@ -0,0 +1,184 @@
...
+     */
+    private function persist($data)
+    {
+        foreach ($data as $scorer) {
+            $dbScorer = $this->objectManager->getRepository(Scorer::class)->findOneBy(['username' => $scorer['username']]);
+            if (!$this->input->hasParameterOption(self::OPTION_OVERWRITE)) {
+                if (!($dbScorer instanceOf Scorer)) {
```
Comment (peetya) [inline src/Command/ImportScorerCommand.php:154]: if the Scorer is not exists in the db, the `findOneBy` will return null, so it's more clear if you check that the value of it is null or not. E.g.:
```
if (!$dbScorer) {
  $this->addNewScorer($scorer);
}
```

File: src/Command/ImportScorerCommand.php [hunk 27]
```diff
@@ -0,0 +1,184 @@
...
+            $this->symfonyStyle->success('import done successfully !!');
+        } else {
+            $this->symfonyStyle->block('try import done successfully !!', 'info', 'fg=black;bg=blue', ' ', true);
+        }
+    }
+
+    /**
+     * persist a new scorer
+     *
+     * @param $scorer
```
Comment (peetya) [inline src/Command/ImportScorerCommand.php]: Type declaration is missing.

File: src/Command/ImportScorerCommand.php [hunk 28]
```diff
@@ -0,0 +1,184 @@
...
+     */
+    private function addNewScorer($scorer)
+    {
+        $scorerEntity = new Scorer();
+        $scorerEntity->setUsername($scorer['username']);
+        $scorerEntity->setPassword($this->encoder->encodePassword($scorerEntity, $scorer['password']));
+        $this->objectManager->persist($scorerEntity);
+    }
+
+    private function updateScorer($scorerEntity, $scorer)
```
Comment (peetya) [inline src/Command/ImportScorerCommand.php:196]: PHPDoc is missing.

File: src/Command/ImportScorerCommand.php [hunk 29]
```diff
@@ -0,0 +1,184 @@
...
+    {
+        $scorerEntity->setPassword($this->encoder->encodePassword($scorerEntity, $scorer['password']));
+        $this->objectManager->persist($scorerEntity);
+    }
+
+    /**
+     * convert and show StopWatch data
+     *
+     * @param $memory
+     * @param $duration
```
Comment (peetya) [inline src/Command/ImportScorerCommand.php:206]: Type declaration is missing for these parameters.

File: src/Command/ImportScorerCommand.php [hunk 30]
```diff
@@ -0,0 +1,184 @@
...
+     */
+    private function showStopwatchData($memory, $duration)
+    {
+        $data['memory'] = round($memory / 1000000, 2);
+        $data['duration'] = round($duration / 1000, 2);
```
Comment (peetya) [inline src/Command/ImportScorerCommand.php:211]: I would say it's not required to create variables of these values. You can just simply pass them to the `writeln`, or to make it more readable, you can use `sprintf`.

File: tests/Unit/Command/ImportScorerCommandTest.php [hunk 31]
```diff
@@ -0,0 +1,158 @@
...
+ * @author antoinep@taotesting.com
+ * @license See LICENCE.md
+ */
+namespace App\tests\Unit\Command;
+
+use Symfony\Bundle\FrameworkBundle\Console\Application;
+use Symfony\Bundle\FrameworkBundle\Test\KernelTestCase;
+use Symfony\Component\Console\Tester\CommandTester;
+
+class ImportScorerCommandTest extends KernelTestCase
```
Comment (peetya) [inline tests/Unit/Command/ImportScorerCommandTest.php]: This is an Integration test and not a Unit test.

File: src/Command/ImportScorerCommand.php [hunk 32]
```diff
@@ -0,0 +1,184 @@
...
+        $this->symfonyStyle->title('Human-Scoring Import');
+        if (!$input->hasParameterOption(self::OPTION_OVERWRITE)) {
+            $output->writeln(['if you want overwrite scorer data in database use flag \'--overwrite\'', '']);
+        }
+        try {
+            $data = $this->dataProvider->getIterator($this->path);
+            $this->persist($data);
+            $this->flush();
+        } catch (\Exception $e) {
+            echo $e->getMessage();
```
Comment (peetya) [inline src/Command/ImportScorerCommand.php]: Use the `OutputInterface` to display something in the console instead of `echo`.

File: tests/Unit/Command/ImportScorerCommandTest.php [hunk 33]
```diff
@@ -0,0 +1,158 @@
...
+        $commandTester = new CommandTester($command);
+        $result = $commandTester->execute(
+            [
+                'path' => null,
+            ],
+            [
+                'capture_stderr_separately' => true
+            ]
+        );
+        $this->assertEquals(1, $result);
```
Comment (peetya) [inline tests/Unit/Command/ImportScorerCommandTest.php]: It would be better to assert the Exception and the Exception message.

File: tests/Unit/Command/ImportScorerCommandTest.php [hunk 34]
```diff
@@ -0,0 +1,158 @@
...
+        $commandTester = new CommandTester($command);
+        $result = $commandTester->execute(
+            [
+                'path' => __DIR__ . '/../../Samples/wrongPath.csv',
+            ],
+            [
+                'capture_stderr_separately' => true
+            ]
+        );
+        $this->assertEquals(1, $result);
```
Comment (peetya) [inline tests/Unit/Command/ImportScorerCommandTest.php]: It would be better to assert the Exception and the Exception message.

File: tests/Unit/Command/ImportScorerCommandTest.php [hunk 35]
```diff
@@ -0,0 +1,158 @@
...
+        $commandTester = new CommandTester($command);
+        $result = $commandTester->execute(
+            [
+                'path' => __DIR__ . '/../../Samples/wrongHeader.csv',
+            ],
+            [
+                'capture_stderr_separately' => true
+            ]
+        );
+        $this->assertEquals(1, $result);
```
Comment (peetya) [inline tests/Unit/Command/ImportScorerCommandTest.php]: It would be better to assert the Exception and the Exception message.

**Review by peetya — CHANGES_REQUESTED** [review]

File: config/services.yaml [hunk 36]
```diff
@@ -25,3 +25,6 @@ services:
 
     # add more service definitions when explicit configuration is needed
     # please note that last definitions always *replace* previous ones
+
+    App\Command\Adapter\AdapterInterface:
```
Comment (gyszucs) [inline config/services.yaml:29]: FYI, until you have only one implementation of an interface, you don't have to declare the implementation, Symfony can recognise it automatically; of course it is not an issue, **not required to be changed**. Btw, an other shorter format to declare this could be `App\Command\Adapter\AdapterInterface: '@App\Command\Adapter\Csv'`

File: src/Command/Adapter/AdapterInterface.php (hunk 21 above)
Comment (gyszucs) [reply src/Command/Adapter/AdapterInterface.php:19]: As I mentioned in the global comment, please start using strict PHP.

File: src/Command/ImportScorerCommand.php [hunk 37]
```diff
@@ -0,0 +1,184 @@
...
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Input\InputArgument;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Console\Style\SymfonyStyle;
+use Symfony\Component\Security\Core\Encoder\UserPasswordEncoderInterface;
+use Symfony\Component\Stopwatch\Stopwatch;
+
+class ImportScorerCommand extends Command
```
Comment (gyszucs) [inline src/Command/ImportScorerCommand.php:23]: This command contains more logic than it should, breaks [Single responsibility principle](https://en.wikipedia.org/wiki/Single_responsibility_principle). Please consider extracting some functionality (mainly the persisting logic) into a separated service.

File: src/Command/ImportScorerCommand.php [hunk 38]
```diff
@@ -0,0 +1,184 @@
...
+     *
+     * @param InputInterface $input
+     * @param OutputInterface $output
+     * @return int
+     */
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $this->stopwatch->start('import');
+        $this->symfonyStyle->title('Human-Scoring Import');
+        if (!$input->hasParameterOption(self::OPTION_OVERWRITE)) {
```
Comment (gyszucs) [inline src/Command/ImportScorerCommand.php]: Use `getOption` instead of `hasParameterOption`. It's not reliable as it is stated in its documentation.

File: src/Command/ImportScorerCommand.php [hunk 39]
```diff
@@ -0,0 +1,184 @@
...
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Console\Style\SymfonyStyle;
+use Symfony\Component\Security\Core\Encoder\UserPasswordEncoderInterface;
+use Symfony\Component\Stopwatch\Stopwatch;
+
+class ImportScorerCommand extends Command
+{
+    const OPTION_OVERWRITE = '--overwrite';
```
Comment (gyszucs) [inline src/Command/ImportScorerCommand.php]: Don't use `--` in the value.

File: src/Command/ImportScorerCommand.php [hunk 40]
```diff
@@ -0,0 +1,184 @@
...
+    const OPTION_FORCE = '--force';
```
Comment (gyszucs) [inline src/Command/ImportScorerCommand.php]: Same here

File: src/Command/ImportScorerCommand.php [hunk 41]
```diff
@@ -0,0 +1,184 @@
...
+    /**
+     *  configure options and argument of the command.
+     */
+    protected function configure()
+    {
+        $this
+            ->setDescription('Import scorers data from a CSV file')
+            ->setHelp('CSV file HEADERS: \'username,password\'');
+        $this->addOption(
+            'force',
```
Comment (gyszucs) [inline src/Command/ImportScorerCommand.php]: Please use the constant (OPTION_FORCE)  here.

File: src/Command/ImportScorerCommand.php [hunk 42]
```diff
@@ -0,0 +1,184 @@
...
+            'f',
+            InputOption::VALUE_NONE,
+            'Causes data ingestion to be applied into storage'
+        );
+        $this->addOption(
+          'overwrite',
```
Comment (gyszucs) [inline src/Command/ImportScorerCommand.php]: Please use the constant (OPTION_OVERWRITE)  here.

File: src/Command/ImportScorerCommand.php [hunk 43]
```diff
@@ -0,0 +1,184 @@
...
+            'o',
+            InputOption::VALUE_NONE,
+            'overwrite Scorer in database'
+        );
+        $this->addArgument(
+            'path',
```
Comment (gyszucs) [inline src/Command/ImportScorerCommand.php]: Please use the constant (ARGUMENT_PATH) here.

File: src/Command/ImportScorerCommand.php [hunk 44]
```diff
@@ -0,0 +1,184 @@
...
+     * if there is option '--overwrite' persist
+     * persist scorers
+     *
+     * @param $data
+     */
+    private function persist($data)
+    {
+        foreach ($data as $scorer) {
+            $dbScorer = $this->objectManager->getRepository(Scorer::class)->findOneBy(['username' => $scorer['username']]);
+            if (!$this->input->hasParameterOption(self::OPTION_OVERWRITE)) {
```
Comment (gyszucs) [inline src/Command/ImportScorerCommand.php]: Use `getOption` instead of `hasParameterOption`. It's not reliable as it is stated in its documentation.

File: src/Command/ImportScorerCommand.php [hunk 45]
```diff
@@ -0,0 +1,184 @@
...
+            }
+        }
+    }
+
+    /**
+     * if there is '--force' option, flush the data in the database
+     */
+    private function flush()
+    {
+        if ($this->input->hasParameterOption(self::OPTION_FORCE)) {
```
Comment (gyszucs) [inline src/Command/ImportScorerCommand.php]: Use `getOption` instead of `hasParameterOption`. It's not reliable as it is stated in its documentation.

File: src/Command/ImportScorerCommand.php (hunk 39 above)
Comment (gyszucs) [reply src/Command/ImportScorerCommand.php]: Why is it needed anyway? What is its purpose?

File: src/Command/ImportScorerCommand.php (hunk 26 above)
Comment (gyszucs) [reply src/Command/ImportScorerCommand.php:154]: And please try to avoid nesting if cases, its hard to follow. You can check some other clean code instructions also: https://github.com/jupeter/clean-code-php#avoid-nesting-too-deeply-and-return-early-part-1

File: src/Command/ImportScorerCommand.php [hunk 46]
```diff
@@ -0,0 +1,184 @@
...
+        $this->objectManager->persist($scorerEntity);
+    }
+
+    /**
+     * convert and show StopWatch data
+     *
+     * @param $memory
+     * @param $duration
+     */
+    private function showStopwatchData($memory, $duration)
```
Comment (gyszucs) [inline src/Command/ImportScorerCommand.php:208]: Probably this function could be extracted into a trait so later you can reuse it in other commands also like we did here: https://github.com/oat-sa/depp-teacher-portal/blob/master/src/Command/StopWatchTrait.php

**Review by gyszucs — CHANGES_REQUESTED** [review]
There is an error during migration:
- `Migration 20190305132526 failed during Execution. SQLSTATE[42S01]: Base table or view already exists: 1050 Table 'scorer' already exists`
It's probably because the table is already created by `bin/console doctrine:schema:create`, so please use defensive migrations like here: https://github.com/oat-sa/depp-teacher-portal/blob/master/src/Migrations/Version20180828093044.php#L14

- Is it intended that the db is limited to only MySQL? I would suggest to support PostgreSQL also.
- Please start using strict PHP https://www.php.net/manual/en/functions.arguments.php#functions.arguments.type-declaration.strict, so make declaration in each PHP and use proper type-hints for arguments and return values.
- Shouldn't we put this repo under `oat-sa`?

As a global review, I can see you are gaining more and more useful knowledge about Symfony, just keep it up. ;)

File: tests/Unit/Command/ImportScorerCommandTest.php (hunk 33 above)
Comment (sirpa44) [reply tests/Unit/Command/ImportScorerCommandTest.php]: the execute() method return an INT and i can't assert the Exception

File: tests/Unit/Command/ImportScorerCommandTest.php (hunk 34 above)
Comment (sirpa44) [reply tests/Unit/Command/ImportScorerCommandTest.php]: the execute() method return an INT and i can't assert the Exception

File: tests/Unit/Command/ImportScorerCommandTest.php (hunk 35 above)
Comment (sirpa44) [reply tests/Unit/Command/ImportScorerCommandTest.php]: the execute() method return an INT and i can't assert the Exception

File: src/Command/Adapter/AdapterInterface.php [hunk 47]
```diff
@@ -0,0 +1,20 @@
...
+/**
+ * Human Scoring Software
+ *
+ * @author antoinep@taotesting.com
+ * @license See LICENCE.md
+ */
+namespace App\Command\Adapter;
+
+
+interface AdapterInterface
```
Comment (Babacooll) [inline src/Command/Adapter/AdapterInterface.php:11]: I think we can find a better name than AdapterInterface. Usually you would prefix this with something as AdapterInterface remains quite vague. Also, in your command, when you inject an AdapterInterface you name it "$dataProvider", so might be worth taking a look.

File: src/Command/ImportScorerCommand.php [hunk 48]
```diff
@@ -0,0 +1,184 @@
...
+use Symfony\Component\Security\Core\Encoder\UserPasswordEncoderInterface;
+use Symfony\Component\Stopwatch\Stopwatch;
+
+class ImportScorerCommand extends Command
+{
+    const OPTION_OVERWRITE = '--overwrite';
+    const OPTION_FORCE = '--force';
+    const ARGUMENT_PATH = 'path';
+
+    protected $encoder;
```
Comment (Babacooll) [inline src/Command/ImportScorerCommand.php:32]: As PHP does not support property types in version 7.2 you might want to add PHPDocs here with the type for each property.

File: src/Command/ImportScorerCommand.php [hunk 49]
```diff
@@ -0,0 +1,184 @@
...
+        $this->stopwatch->start('import');
+        $this->symfonyStyle->title('Human-Scoring Import');
+        if (!$input->hasParameterOption(self::OPTION_OVERWRITE)) {
+            $output->writeln(['if you want overwrite scorer data in database use flag \'--overwrite\'', '']);
+        }
+        try {
+            $data = $this->dataProvider->getIterator($this->path);
+            $this->persist($data);
+            $this->flush();
+        } catch (\Exception $e) {
```
Comment (Babacooll) [inline src/Command/ImportScorerCommand.php]: Please catch \Throwable instead of \Exception as it is more low-level

File: src/Command/ImportScorerCommand.php [hunk 50]
```diff
@@ -0,0 +1,184 @@
...
+    }
+
+
+    /**
+     * if there is option '--overwrite' persist
+     * persist scorers
+     *
+     * @param $data
+     */
+    private function persist($data)
```
Comment (Babacooll) [inline src/Command/ImportScorerCommand.php:149]: Missing type here, you could typehint with _iterable_ or _array_

File: src/Command/ImportScorerCommand.php (hunk 28 above)
Comment (Babacooll) [reply src/Command/ImportScorerCommand.php:196]: + argument typing

File: tests/Integration/Command/ImportScorerCommandTest.php [hunk 51]
```diff
@@ -0,0 +1,158 @@
...
+        $commandTester = new CommandTester($command);
+        $result = $commandTester->execute(
+            [
+                'path' => __DIR__ . '/../../Samples/wrongPath.csv',
+            ],
+            [
+                'capture_stderr_separately' => true
+            ]
+        );
+        $this->assertEquals(1, $result);
```
Comment (Babacooll) [inline tests/Integration/Command/ImportScorerCommandTest.php:64]: On each of the failure tests, you might also want to ensure that no data has been modified in your database.

**Review by Babacooll — CHANGES_REQUESTED** [review]

File: config/services.yaml (hunk 36 above)
Comment (sirpa44) [reply config/services.yaml:29]: ok !!good to know, thanks for the information :)

**Comment by sirpa44** [comment]
now on oat-sa/tao-human-scoring
